File: Projects/Clock/Clock/Clock.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Clock
{
    public class Clock
    {
        private Counter _hours;
        private Counter _minutes;
        private Counter _seconds;
        private bool _is24HourFormat;

        public Clock()
        {
            Console.WriteLine("Please enter your student ID: ");
            string? input = Console.ReadLine();
            if (string.IsNullOrEmpty(input))
            {
                throw new ArgumentNullException("Student ID cannot be null or empty.");
            }
            int studentID = int.Parse(input);

            int lastDigit = studentID % 10;
            _is24HourFormat = lastDigit > 5;

            _hours = new Counter("hours");
            _minutes = new Counter("minutes");
            _seconds = new Counter("seconds");
        }

        public Clock(bool is24HourFormat)
        {
            this._is24HourFormat = is24HourFormat;
            _hours = new Counter("hours");
            _minutes = new Counter("minutes");
            _seconds = new Counter("seconds");
        }

        public void Tick()
        {
            _seconds.Increment();
            if (_seconds.Ticks >= 60)
            {
                _seconds.Reset();
                _minutes.Increment();
                if (_minutes.Ticks >= 60)
                {
                    _minutes.Reset();
                    _hours.Increment();
                    int maxHours = 24;
                    if (_hours.Ticks >= maxHours)
                    {
                        _hours.Reset();
                    }
                }
            }
        }

        public void Reset()
        {
            _hours.Reset();
            _minutes.Reset();
            _seconds.Reset();
        }

        public string Time
        {
            get
                {
                if (_is24HourFormat)
                {
                    return ($"{_hours.Ticks:D2}:{_minutes.Ticks:D2}:{_seconds.Ticks:D2}");
                }
                else
                {
                    int displayHours = _hours.Ticks % 12;
                    // Convert 0 hours to 12 for 12-hour format
                    if (displayHours == 0) displayHours = 12; 
                    string period = _hours.Ticks < 12 ? "AM" : "PM";
                    return ($"{displayHours:D2}:{_minutes.Ticks:D2}:{_seconds.Ticks:D2} {period}");
                }
            }
        }

        public int Hours
        {
            get
            {
                return _hours.Ticks;
            }
        }

        public int Minutes
        {
            get
            {
                return _minutes.Ticks;
            }
        }

        public int Seconds
        {
            get
            {
                return _seconds.Ticks;
            }
        }
    }
}

File: Projects/ClockPython/Clock.py
from Counter import Counter

class Clock:
    def __init__(self, is_24_hour_format=None):
        if is_24_hour_format is None:
            student_id = input("Please enter your student ID: ")
            if not student_id:
                raise ValueError("Student ID cannot be null or empty.")
            last_digit = int(student_id) % 10
            self._is_24_hour_format = last_digit > 5
        else:
            self._is_24_hour_format = is_24_hour_format

        self._hours = Counter("hours")
        self._minutes = Counter("minutes")
        self._seconds = Counter("seconds")

    def Tick(self):
        self._seconds.Increment()
        if self._seconds.Ticks >= 60:
            self._seconds.Reset()
            self._minutes.Increment()
            if self._minutes.Ticks >= 60:
                self._minutes.Reset()
                self._hours.Increment()
                max_hours = 24
                if self._hours.Ticks >= max_hours:
                    self._hours.Reset()

    def Reset(self):
        self._hours.Reset()
        self._minutes.Reset()
        self._seconds.Reset()

    @property
    def Time(self):
        if self._is_24_hour_format:
            return f"{self._hours.Ticks:02d}:{self._minutes.Ticks:02d}:{self._seconds.Ticks:02d}"
        else:
            display_hours = self._hours.Ticks % 12
            if display_hours == 0:
                display_hours = 12
            period = "AM" if self._hours.Ticks < 12 else "PM"
            return f"{display_hours:02d}:{self._minutes.Ticks:02d}:{self._seconds.Ticks:02d} {period}"

    @property
    def Hours(self):
        return self._hours.Ticks

    @property
    def Minutes(self):
        return self._minutes.Ticks

    @property
    def Seconds(self):
        return self._seconds.Ticks
    
